File: src-tauri/src/models/audio_detection.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum AudioDetectionStatus {
    Pending,
    Processing,
    Completed,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TimestampSegment {
    pub start_seconds: f64,
    pub end_seconds: f64,
    pub confidence: f64, 
    pub label: String,   
}

/// Result of audio event detection inference
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioDetectionResult {
    pub segments: Vec<TimestampSegment>,
    pub total_duration_seconds: f64,
    pub detected_duration_seconds: f64,
    pub model_version: String,
}

/// Audio detection job status and metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AudioDetectionJob {
    pub id: String,
    pub input_file: String,
    pub status: AudioDetectionStatus,
    pub progress: u8, 
    pub stage: Option<String>,
    pub created_at: String,
    pub completed_at: Option<String>,
    pub error: Option<String>,
    pub result: Option<AudioDetectionResult>,
}

impl AudioDetectionJob {
    pub fn new(id: String, input_file: String) -> Self {
        Self {
            id,
            input_file,
            status: AudioDetectionStatus::Pending,
            progress: 0,
            stage: None,
            created_at: chrono::Utc::now().to_rfc3339(),
            completed_at: None,
            error: None,
            result: None,
        }
    }
}

/// Model configuration for inference
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelConfig {
    pub model_path: Option<String>,
    pub window_size_ms: u32,
    pub hop_size_ms: u32,
    pub confidence_threshold: f64,
    pub min_segment_duration_ms: u32,
    pub merge_gap_ms: u32,
}

impl Default for ModelConfig {
    fn default() -> Self {
        Self {
            model_path: None,
            window_size_ms: 1000,
            hop_size_ms: 250,
            confidence_threshold: 0.7,
            min_segment_duration_ms: 500,
            merge_gap_ms: 300,
        }
    }
}

/// Training sample entry for manifest
#[derive(Debug, Clone, Serialize, Deserialize)]
#[allow(dead_code)] 
pub struct TrainingSample {
    pub file: String,
    pub label: String, 
    pub source: String,
}

/// Training data manifest
#[derive(Debug, Clone, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct TrainingManifest {
    pub positive_samples: Vec<TrainingSample>,
    pub negative_samples: Vec<TrainingSample>,
    pub hard_negative_samples: Vec<TrainingSample>,
}

// ============================================================================
// Enhance Model Mode Types
// ============================================================================

/// A single feedback sample from user labeling
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeedbackSample {
    pub id: String,
    pub source_file: String,
    pub start_seconds: f64,
    pub end_seconds: f64,
    pub original_confidence: f64,
    pub user_label: String, // "correct" or "wrong"
    pub is_manual: bool,
    pub created_at: String,
}

/// A manually-marked segment (false negative)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ManualSegment {
    pub id: String,
    pub start_seconds: f64,
    pub end_seconds: f64,
    pub created_at: String,
}

/// A feedback session containing all corrections for a detection job
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeedbackSession {
    pub id: String,
    pub source_file: String,
    pub job_id: String,
    pub model_version: String,
    pub samples: Vec<FeedbackSample>,
    pub manual_positives: Vec<ManualSegment>,
    pub created_at: String,
    pub updated_at: String,
}

/// UI-facing training configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UITrainingConfig {
    pub epochs: u32,
    pub learning_rate: f64,
    #[serde(default)]
    pub bulk_positive_files: Vec<String>,
    #[serde(default = "default_fine_tune")]
    pub fine_tune: bool,
    #[serde(default = "default_freeze_layers")]
    pub freeze_layers: bool,
    #[serde(default = "default_unfreeze_after")]
    pub unfreeze_after: u32,
}

fn default_fine_tune() -> bool {
    true
}

fn default_freeze_layers() -> bool {
    true
}

fn default_unfreeze_after() -> u32 {
    5
}

impl Default for UITrainingConfig {
    fn default() -> Self {
        Self {
            epochs: 15,           // Reduced from 30 for fine-tuning
            learning_rate: 0.0001, // Reduced from 0.001 for fine-tuning
            bulk_positive_files: vec![],
            fine_tune: true,
            freeze_layers: true,
            unfreeze_after: 5,
        }
    }
}

File: src/types/audioDetection.ts
export type AudioDetectionStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'cancelled';

export interface TimestampSegment {
  start_seconds: number;
  end_seconds: number;
  confidence: number; // 0.0 - 1.0
  label: 'target_audio';
}

export interface AudioDetectionResult {
  segments: TimestampSegment[];
  total_duration_seconds: number;
  detected_duration_seconds: number;
  model_version: string;
}

export interface AudioDetectionJob {
  id: string;
  input_file: string;
  status: AudioDetectionStatus;
  progress: number;
  stage: string | null;
  created_at: string;
  completed_at: string | null;
  error: string | null;
  result: AudioDetectionResult | null;
}

export interface ModelConfig {
  model_path: string | null; 
  window_size_ms: number; 
  hop_size_ms: number; 
  confidence_threshold: number; 
  min_segment_duration_ms: number; 
  merge_gap_ms: number; 
}

export const DEFAULT_MODEL_CONFIG: ModelConfig = {
  model_path: null,
  window_size_ms: 1000,
  hop_size_ms: 250,
  confidence_threshold: 0.7,
  min_segment_duration_ms: 500,
  merge_gap_ms: 300,
};

// Command parameters
export interface SubmitAudioDetectionParams {
  input_file: string;
  config: ModelConfig | null; 
}

export interface SubmitAudioDetectionResult {
  job_id: string;
}

// Event payloads
export interface AudioDetectionProgressEvent {
  job_id: string;
  percent: number;
  stage: string;
}

export interface AudioDetectionCompletedEvent {
  job_id: string;
  result: AudioDetectionResult;
}

export interface AudioDetectionFailedEvent {
  job_id: string;
  error: string;
}

// Training data manifest types 
export interface TrainingSample {
  file: string;
  label: 'target_audio' | 'other';
  source: string; 
}

export interface TrainingManifest {
  positive_samples: TrainingSample[];
  negative_samples: TrainingSample[];
  hard_negative_samples: TrainingSample[];
}

// Training configuration
export interface TrainingConfig {
  batch_size: number;
  learning_rate: number;
  weight_decay: number;
  epochs: number;
  early_stopping_patience: number;
  augmentation: {
    time_mask_max_width: number;
    freq_mask_max_width: number;
    gain_range: [number, number];
    mixup_alpha: number;
  };
  class_weights: {
    positive: number;
    negative: number;
  };
  output_dir: string;
  export_onnx: boolean;
}

// ============================================================================
// Enhance Model Mode Types
// ============================================================================

// Feedback types for user corrections on inference results
export interface FeedbackSample {
  id: string;
  source_file: string;
  start_seconds: number;
  end_seconds: number;
  original_confidence: number;
  user_label: 'correct' | 'wrong';
  is_manual: boolean; // true for user-added false negatives
  created_at: string;
}

// Manual segment for false negatives (segments the model missed)
export interface ManualSegment {
  id: string;
  start_seconds: number;
  end_seconds: number;
  created_at: string;
}

export interface FeedbackSession {
  id: string;
  source_file: string;
  job_id: string;
  model_version: string;
  samples: FeedbackSample[];
  manual_positives: ManualSegment[]; // User-marked false negatives
  created_at: string;
  updated_at: string;
}

// UI-facing training configuration (simplified)
export interface UITrainingConfig {
  epochs: number;
  learning_rate: number;
  bulk_positive_files?: string[]; // Full audio files to slice as positive samples
  bulk_negative_files?: string[]; // Full audio files to slice as negative samples
  fine_tune?: boolean;           // Default: true - fine-tune existing model instead of training from scratch
  freeze_layers?: boolean;       // Default: true - freeze early conv layers to prevent forgetting
  unfreeze_after?: number;       // Default: 5 - unfreeze all layers after N epochs for gradual fine-tuning
}

export const DEFAULT_UI_TRAINING_CONFIG: UITrainingConfig = {
  epochs: 15,             // Reduced from 30 for fine-tuning
  learning_rate: 0.0001,  // Reduced from 0.001 for fine-tuning
  bulk_positive_files: [],
  bulk_negative_files: [],
  fine_tune: true,
  freeze_layers: true,
  unfreeze_after: 5,
};

// Training progress state
export interface TrainingProgress {
  percent: number;
  epoch: number;
  total_epochs: number;
  stage: string;
  metrics: {
    train_loss: number;
    val_loss: number;
    val_f1: number;
    val_accuracy: number;
  } | null;
}

// Event payloads for training
export interface TrainingProgressEvent {
  percent: number;
  stage: string;
  metrics: TrainingProgress['metrics'];
}

export interface TrainingCompleteEvent {
  success: boolean;
  model_path: string;
  final_metrics: TrainingProgress['metrics'];
  samples_used: number;
}
